### PR TITLE
[ARM] Stop gluing 1-bit shifts

### PR DESCRIPTION
**Pull request by s-barannikov**
1. When two (or more) nodes are glued, DAG scheduler will always schedule them as one piece, i.e. it will not allow any instructions to be scheduled between them. It does so because if nodes are glued this usually means that there is an implicit register dependency between them, and an intervening node could clobber this physical register. When emitting such nodes into machine IR, they will also be stuck together, e.g.:
```
    %9:gpr = MOVsrl_glue killed %8, implicit-def $cpsr
    %10:gpr = RRX %3, implicit $cpsr
```

2. If a node has Glue result, SelectionDAG will not try to CSE this node. If it did, it would break the implicit physical register dependency. In practice this means that if a node with Glue result has multiple uses, it has to be duplicated before each use. This the reason for `ARMTargetLowering::duplicateCmp` to exist.

When using normal data dependency, dependent nodes can freely be scheduled around. If there is a physical register dependency between nodes, the physical register will be copied to/from a virtual register, allowing other nodes to intervene between them. The resulting machine IR might look like this:
```
    %9:gpr = LSRs1 killed %8, implicit-def $cpsr
    %10:gpr = COPY $cpsr
    %11:gpr = ORRrsi killed %9, %3, 242, 14 /* CC::al */, $noreg, $noreg
    %12:gpr = BICri killed %11, -2147483648, 14 /* CC::al */, $noreg, $noreg
    $cpsr = COPY %10
    %13:gpr = RRX %3, implicit $cpsr
```

The two copies are likely to be eliminated by register coalescer, given that there are no instructions between them that clobber this physical register. If the copies are unwanted in the first place (they could be expensive or impossible), DAG scheduler will try to avoid inserting them wherever possible, and the resulting machine IR will look like this:
```
    %9:gpr = LSRs1 killed %8, implicit-def $cpsr
    %10:gpr = ORRrsi killed %9, %3, 242, 14 /* CC::al */, $noreg, $noreg
    %11:gpr = BICri killed %10, -2147483648, 14 /* CC::al */, $noreg, $noreg
    %12:gpr = RRX %3, implicit $cpsr
```

On ARM, arithmetic operations and LSLS already use the new data flow approach. This patch extends it to include 1-bit shifts.

### CODE REVIEW

File: llvm/lib/Target/ARM/ARMInstrInfo.td [hunk 1]
```diff
@@ -191,9 +206,9 @@ def ARMasrl          : SDNode<"ARMISD::ASRL", SDT_ARMIntShiftParts, []>;
 def ARMlsrl          : SDNode<"ARMISD::LSRL", SDT_ARMIntShiftParts, []>;
 def ARMlsll          : SDNode<"ARMISD::LSLL", SDT_ARMIntShiftParts, []>;
 
-def ARMsrl_glue      : SDNode<"ARMISD::SRL_GLUE", SDTIntUnaryOp, [SDNPOutGlue]>;
-def ARMsra_glue      : SDNode<"ARMISD::SRA_GLUE", SDTIntUnaryOp, [SDNPOutGlue]>;
-def ARMrrx           : SDNode<"ARMISD::RRX"     , SDTIntUnaryOp, [SDNPInGlue ]>;
+def ARMsrl_glue : SDNode<"ARMISD::SRL_GLUE", SDTIntUnaryOpWithFlagsOut>;
```
Comment (topperc) [inline llvm/lib/Target/ARM/ARMInstrInfo.td:209]: Drop glue from the name?

**Comment by statham-arm** [comment]
Can you expand the commit message to explain the benefit of this change? It says _what_ you're doing, but expects the reader to already understand _why_.

I haven't needed to get involved with "glue" in SelectionDAG before. But it looks as if the idea is that now when we generate an LSRS + RRX pair, they're tied together by a precise data flow dependency saying the second instruction uses the CPSR written by the first, whereas previously the "glue" system said something more approximate like "just don't separate this pair of instructions"? So now you could imagine a backend pass separating the two instructions, if something else could safely be put in between.

But I don't see any examples in the changed tests where that's actually happened. All the test changes look as if they've just perturbed the register allocation.

Is this change intended to improve performance? Or is there some other purpose?

**Comment by s-barannikov** [comment]
@statham-arm I've updated the description, PTAL.

> But I don't see any examples in the changed tests where that's actually happened. All the test changes look as if they've just perturbed the register allocation.

This is probably due to 1-bit shift optimization not triggering often on existing tests. In the next patch I change FP comparisons to use data flow too, and there is a significant difference. Here is just a small example:
image

I didn't submit it yet because it depends on this PR.

**Comment by s-barannikov** [comment]
> I didn't submit it yet because it depends on this PR.

https://github.com/llvm/llvm-project/pull/116676

File: llvm/lib/Target/ARM/ARMScheduleM7.td [hunk 2]
```diff
@@ -325,7 +325,7 @@ def M7Ex1ReadNoFastBypass : SchedReadAdvance<-1, [WriteLd, M7LoadLatency1]>;
 def : InstRW<[WriteALUsi, M7Ex1ReadNoFastBypass, M7Read_ISS],
              (instregex "t2(ADC|ADDS|ADD|BIC|EOR|ORN|ORR|RSBS|RSB|SBC|SUBS)rs$",
                         "t2(SUB|CMP|CMNz|TEQ|TST)rs$",
-                        "t2MOVsr(a|l)")>;
+                        "t2(A|L)SRs1")>;
```
Comment (statham-arm) [inline llvm/lib/Target/ARM/ARMScheduleM7.td:328]: Tiny nit: should this regex end in `$`, to match the two above it? I don't think you're intentionally matching other opcodes that start with `t2ASRs1` or `t2LSRs1`. (Whereas the previous code _was_ matching things that had extra stuff after `t2MOVsra`/`t2MOVsrl`.

**Review by statham-arm — APPROVED** [review]
Thanks – that's a great explanation and I'm glad I asked! I'd guessed your point 1 (scheduling flexibility), but hadn't spotted point 2 at all (eliminating duplication). So this actually gives _more_ benefit than I'd realised.

LGTM, although I've spotted one last _tiny_ nit.

**Pull request merged**